agent/ebs/watcher.go

```diff
@@ -0,0 +1,212 @@
+	var err error
+	var wg sync.WaitGroup
+	wg.Add(1)
+	w.mailbox <- func() {
+		defer wg.Done()
+		empty := len(w.agentState.GetAllPendingEBSAttachments()) == 0
+
+		err := w.handleEBSAttachment(ebs)
+		if err != nil {
+			log.Warnf("Failed to handle resource attachment %v", ebs.String())
+		}
+		if empty && len(w.agentState.GetAllPendingEBSAttachments()) == 1 {
+			w.scanTicker.Stop()
+			w.scanTicker = time.NewTicker(apiebs.ScanPeriod)
+		}
+	}
+	wg.Wait()
+	return err
+}
+
+// handleEBSAttachment will handle an EBS attachment via the following:
+// 1. Check whether we already have this attachment in state, if so, return
+// 2. Otherwise add the attachment to state, start its ack timer, and save the state
+func (w *EBSWatcher) handleEBSAttachment(ebs *apiebs.ResourceAttachment) error {
+	if ebs.AttachmentProperties[apiebs.ResourceTypeName] != apiebs.ElasticBlockStorage {
+		log.Warn("Resource type not Elastic Block Storage. Skip handling resource attachment.")
+		return nil
+	}
+	volumeID := ebs.AttachmentProperties[apiebs.VolumeIdName]
+	_, ok := w.agentState.GetEBSByVolumeId(volumeID)
+	if ok {
+		log.Infof("EBS Volume attachment already exists. Skip handling EBS attachment %v.", ebs.String())
+		return nil
+	}
+
+	if err := w.addEBSAttachmentToState(ebs); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *EBSWatcher) NotifyFound(foundVolumes []string) {
+	for _, volumeId := range foundVolumes {
+		w.notifyFoundEBS((volumeId))
+	}
+}
+
+// notifyFoundEBS will mark it as found within the agent state
+func (w *EBSWatcher) notifyFoundEBS(volumeId string) {
+	w.mailbox <- func() {
+		ebs, ok := w.agentState.GetEBSByVolumeId(volumeId)
+		if !ok {
+			log.Warnf("Unable to find EBS volume with volume ID: %v.", volumeId)
+			return
+		}
+
+		if ebs.HasExpired() {
+			log.Warnf("EBS status expired, no longer tracking EBS volume: %v.", ebs.String())
+			return
+		}
+
+		if ebs.IsSent() {
+			log.Warnf("State change event has already been emitted for EBS volume: %v.", ebs.String())
+			return
+		}
+
+		if ebs.IsAttached() {
+			log.Infof("EBS volume: %v, has already been found. Skip notifying EBS volume.", ebs.String())
+			return
+		}
+
+		ebs.StopAckTimer()
+		ebs.SetAttachedStatus()
+
+		log.Infof("Successfully found attached EBS volume: %v", ebs.String())
+		if len(w.agentState.GetAllPendingEBSAttachments()) == 0 {
+			log.Info("No more attachments to scan for. Stopping scan ticker.")
+			w.scanTicker.Stop()
+		}
+	}
+}
+
+// RemoveAttachment will stop tracking an EBS attachment
+func (w *EBSWatcher) RemoveAttachment(volumeID string) {
+	w.mailbox <- func() {
+		w.removeEBSAttachment(volumeID)
+	}
+}
+
+func (w *EBSWatcher) removeEBSAttachment(volumeID string) {
+	// TODO: Remove the EBS volume from the data client.
+	w.agentState.RemoveEBSAttachment(volumeID)
+	if len(w.agentState.GetAllPendingEBSAttachments()) == 0 {
+		log.Info("No more attachments to scan for. Stopping scan ticker.")
+		w.scanTicker.Stop()
```

Comment:
Why do we need to stop the ticker at so many places? I am concerned that the complexity of ticker state management makes it easy to introduce bugs.

Reply:
We want to conservatively start scanning only when necessary. Thus, we should stop the ticker whenever we no longer have any pending attachments. This happens when:
1. We finished scanning volumes
2. We removed a volume

There's also the case where the context gets canceled and so we'd want to stop the ticker as well.